### PR TITLE
feat: add delegate borrowing feature

### PR DESCRIPTION
**Pull request by kkirka**
This PR adds delegated borrowing feature. It would allow users to delegate their borrowing power to some account. In `DelegateBorrowers/` directory there is an example delegate – it only allows to swap one debt position to another, e.g. by repaying BUSD and borrowing ETH. The conversion is done based on the current oracle price.

The Comptroller contract is currently oversize by 94 bytes, tests are not yet implemented.

### CODE REVIEW

**Comment by kkirka** [comment]
Initially, I tried creating `addDelegate` and `removeDelegate` functions but it made the Comptroller oversize by 200+ bytes, which isn't acceptable. I merged these into `updateDelegate`, which is a bit less intuitive, but should work 🤷

**Review by chechu — APPROVED** [review]

File: contracts/DelegateBorrowers/SwapDebtDelegate.sol [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    /**
+     * @notice Repays a borrow in repayTo.underlying() and borrows borrowFrom.underlying()
+     * @param borrower The address of the borrower, whose debt to swap
+     * @param repayTo VToken to repay the debt to
+     * @param borrowFrom VToken to borrow from
+     * @param repayAmount The amount to repay in terms of repayTo.underlying()
+     */
+    function swapDebt(
```
Comment (narayanprusty) [inline contracts/DelegateBorrowers/SwapDebtDelegate.sol:57]: I think we should emit an event when debt is swapped successfully

File: contracts/DelegateBorrowers/SwapDebtDelegate.sol [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        uint256 actualRepaymentAmount = _repay(repayTo, borrower, repayAmount);
+        uint256 amountToBorrow = _convert(repayTo, borrowFrom, actualRepaymentAmount);
+        _borrow(borrowFrom, borrower, amountToBorrow);
+    }
+
+    /**
+     * @notice Transfers tokens, accidentially sent to this contract, to the owner
+     * @param token ERC-20 token to sweep
+     */
+    function sweepTokens(IERC20Upgradeable token) external onlyOwner {
```
Comment (narayanprusty) [inline contracts/DelegateBorrowers/SwapDebtDelegate.sol:72]: Here also we should emit an event when tokens are swept successfully

**Review by narayanprusty — APPROVED** [review]

**Review by defcon022 — APPROVED** [review]

**Review by UrAvgDeveloper — APPROVED** [review]

**Review by chechu — APPROVED** [review]
LGTM

**Pull request merged**